packages/logging/src/logging-and-telemetry.ts

```diff
@@ -82,30 +103,72 @@ class LoggingAndTelemetry implements MongoshLoggingAndTelemetry {
   private readonly mongoshVersion: string;
 
   private log: MongoLogWriter;
-  private pendingLogEvents: CallableFunction[] = [];
+  private pendingBusEvents: CallableFunction[] = [];
+  private pendingTelemetryEvents: CallableFunction[] = [];
   private isSetup = false;
-  private isBufferingEvents = false;
+  private isBufferingBusEvents = false;
+  private isBufferingTelemetryEvents = false;
+
+  private deviceId: string | undefined;
+  /** @internal */
+  public setupTelemetryPromise: Promise<void> = Promise.resolve();
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  private resolveDeviceId: (value: string) => void = () => {};
 
   constructor({
     bus,
     analytics,
     userTraits,
     mongoshVersion,
+    deviceId,
   }: MongoshLoggingAndTelemetryArguments) {
     this.bus = bus;
     this.analytics = analytics;
     this.log = LoggingAndTelemetry.dummyLogger;
     this.userTraits = userTraits;
     this.mongoshVersion = mongoshVersion;
+    this.deviceId = deviceId;
   }
 
   public setup(): void {
     if (this.isSetup) {
       throw new Error('Setup can only be called once.');
     }
+    this.isBufferingTelemetryEvents = true;
+
+    this.setupTelemetryPromise = this.setupTelemetry();
     this.setupBusEventListeners();
+
     this.isSetup = true;
-    this.isBufferingEvents = true;
+  }
+
+  public flush(): void {
+    // Run any telemetry events even if device ID hasn't been resolved yet
+    this.runAndClearPendingTelemetryEvents();
+
+    // Run any other pending events with the set or dummy log for telemetry purposes.
+    this.runAndClearPendingBusEvents();
+
+    this.resolveDeviceId('unknown');
+  }
+
+  private async setupTelemetry(): Promise<void> {
+    if (!this.deviceId) {
+      try {
+        this.deviceId ??= await Promise.race([
+          getDeviceId(),
+          new Promise<string>((resolve) => {
+            this.resolveDeviceId = resolve;
+          }),
+        ]);
+      } catch (error) {
+        this.bus.emit('mongosh:error', error as Error, 'telemetry');
```

Comment:
do we want to actually log this error? Seems like something we'd like to track ourselves but not necessarily show to users as it's not going to affect anything other than telemetry

Reply:
ah yeah right, I was under the false impression that logging in this case would also be `stdout`-ing the output, nevermind then, agreed